### ISSUE TITLE
Add an option to clear logs periodically

### ISSUE THREAD

**Issue by najjara** [celzero/rethink-app#1178]
Currently the user can either enable logging in order to see connections, or disable it then nothing is displayed at all (not even a live feed).

These logs are kept forever, unless cleared manually by the user.
There should be an option for rotating logs periodically.

There are myriad reasons for this. The logs can reach huge sizes. Exempted apps, such as browsers, are still logged, and this is a big privacy issue (the user's entire history is saved forever if the user doesn't take proactive steps), and wastes battery spent logging unnecessary information. Moreover, the logs are stored in cleartext, which is yet another massive privacy & security problem.

Recommendations:
- Add the option to rotate logs
- Show a feed of connections (e.g. written to /dev/null) since opening the app and while the app is open even when logging is disabled
- Add the option to exempt an app from logging
- Consider encrypting logs, and adding an option for encrypted export

**Comment by ignoramous** [comment]
> These logs are kept forever, unless cleared manually by the user.

Not really. The logs are truncated at 150k (iirc) entries periodically.

> Moreover, the logs are stored in cleartext, which is yet another massive privacy & security problem.

Well, no. The database is stored in Rethink's internal data directory which is inaccessible to ALL other apps. And ALL such internal directories are encrypted with user-dependent credentials starting Android 7+ (see: [Android encryption in depth](https://blog.quarkslab.com/android-data-encryption-in-depth.html) by Quark Labs).

> Show a feed of connections (e.g. written to /dev/null) since opening the app and while the app is open even when logging is disabled

Such functionality is already exposed via pcap (*Configure* -> *Settings* -> *Packet capture* -> *Output to Logcat* / *Output to the Downloads folder*). Does that suffice?

> Add the option to exempt an app from logging

Pending: #765 (it isn't straight-forward, but we'll see).

> Consider encrypting logs, and adding an option for encrypted export

As mentioned, all of the `/data` partition in Android 7+ is encrypted. I don't see much value in encrypting the exported sqlite3 file because it is exported to the "Downloads" folder on the device and so the user has plenty options to encrypt it before pulling that folder out. It is something we can consider, but doesn't seem like it adds that much value for the complexity it'll bring?

**Comment by najjara** [comment]
> The logs are truncated at 150k (iirc) entries periodically.

150k lines is a very long log. I would prefer that logs are cleared every 24 hours.

> Such functionality is already exposed via pcap (Configure -> Settings -> Packet capture -> Output to Logcat / Output to the Downloads folder). Does that suffice?

I can't see a live feed of connections in the app, a la Wireshark, with that option. Is the logging functionality essential to the app's function (meaning it can't whitelist/blacklist when not logging)? 
My scenario is that I don't care about past connections, but I still want to be able to troubleshoot apps that malfunction or misbehave when the app is on, and which lists are causing the issue. For me this can accomplished if temp capturing is active and starts when RDNS is in the foreground (or a new separate function called temp cap) that only captures the most recent connections (for example going back to the app having issues, and seeing in RDNS which connections it's attempting to establish). Such information is not needed after dealing with it, and should be wiped when putting RDNS back in the background.

**Comment by ignoramous** [comment]
> Is the logging functionality essential to the app's function

Yes, for stats and other aggregation like data use. For firewall itself, logging or no has no affect.

> if temp capturing is active and starts when RDNS is in the foreground (or a new separate function called temp cap) that only captures the most recent connections. Such information is not needed after dealing with...

This is a different ask from clearing logs periodically. Which one do you think is more ergonomic? A temp capture of logs (say the recent 1000 entries) or clearing it on an aggressive schedule?

**Comment by LeoPrissberg** [comment]
I'm working on this

**Comment by fluffylenny** [comment]
Hi @ignoramous, I am currently working with @LeoPrissberg on this issue.

If our understanding is correct, we plan to change the code in PurgeConnectionLogs.kt so that the NUMBER_OF_DAYS_TO_PURGE variable is set to a mutable, user-selected number of days. We were thinking we could add an attribute containing the purge interval to the persistent state based on the user’s selection in MiscSettingsActivity.kt.

Just a couple of questions:

1. What should the min/max possible “purge interval” in the settings be, ie. 1 hour, 1 day, 30 days? We were thinking 1hr as the minimum and 30 days as the maximum.

2. Should there be an option for the user to keep all their connection logs without purging at all (at least until the logs reach a certain number of entries), or is it a bad idea efficiency-wise?

3. If the minimum possible purge interval should be less than 4 hours (the current amount of time the PurgeConnectionsLog.kt worker waits in between runs) would it be acceptable to increase the frequency of that worker’s deployment in WorkScheduler.kt to maintain real-time accuracy?

4. Should the event log purge frequency also be made customizable since the PurgeConnectionsLog worker handles the purging of both event and connection logs? We were just wondering since event logs and connection logs have different purge intervals by default at the moment (every 4 days and 7 days, respectively).

**Comment by hussainmohd-a** [comment]
Hi @LeoPrissberg, @fluffylenny Thanks for picking this up.

> 1. What should the min/max possible “purge interval” in the settings be, ie. 1 hour, 1 day, 30 days? We were thinking 1hr as the minimum and 30 days as the maximum.

For now, I'd suggest limiting the configuration to a maximum of 7 days. 

> 2. Should there be an option for the user to keep all their connection logs without purging at all (at least until the logs reach a certain number of entries), or is it just a plain bad idea efficiency-wise?

This can significantly increase the app's storage size, so it's better to always purge the logs for now. 

> 3. If the minimum possible purge interval should be less than 4 hours (the current amount of time the PurgeConnectionsLog.kt worker waits in between runs) would it be acceptable to increase the frequency of that worker’s deployment in WorkScheduler.kt to maintain real-time accuracy?

The worker frequency can be adjusted based on the selected purge interval, for e.g.,

| Configured purge interval | Worker interval |
| ------------------------- | --------------- |
| 1 hour                    | 1 hour          |
| 6 hours                   | 3 hours         |
| 12 hours                  | 4 hours         |
| 1 day                     | 12 hours        |
| 7 days                    | 24 hours        |


> 4. Should the event log purge frequency also be made customizable since the PurgeConnectionsLog worker handles the purging of both event and connection logs? We were just wondering since event logs and connection logs have different purge intervals by default at the moment (every 4 days and 7 days, respectively).

I think we can have a single configuration that controls the purge interval for both the event and connection logs.
